### ISSUE TITLE
Broad research on seven-segment display libs

### ISSUE THREAD

**Issue by vincent-karuri** [opensrp/fhircore#82]
Look into these links (and perhaps others) for clues on what's out there for 7-segment-display capture systems:

1. https://github.com/SachaIZADI/Seven-Segment-OCR
2. https://github.com/alexcohn/tess-two
3. https://www.youtube.com/watch?v=J0jg1Hj2GhU
4. https://code.google.com/archive/p/serak-tesseract-trainer/downloads
5. https://stackoverflow.com/questions/17672705/text-detection-on-seven-segment-display-via-tesseract-ocr
6. https://play.google.com/store/apps/details?id=jp.gr.java_conf.coskx.ddreader2.trial&hl=en_US&gl=US
7. https://github.com/arundil/OCR7Segments/tree/master/OCR7SegmentsApp 
8. https://github.com/arturaugusto/display_ocr (recommended by UW)
9. https://github.com/jiweibo/SSOCR/blob/master/ssocr.py (a single script)

**Comment by vincent-karuri** [comment]
@shoaibmushtaq25 could you indicate which of these you'll be picking up?

**Comment by shoaibmushtaq25** [comment]
Currently, I am working on this one https://github.com/arundil/OCR7Segments/tree/master/OCR7SegmentsApp
Converted old project to androidx and now trying to run it on mobile.

**Comment by vincent-karuri** [comment]
MLKit work is documented here https://github.com/OpenSRP/fhircore/issues/69.

**Comment by shoaibmushtaq25** [comment]
After that, I will be testing number 6 - https://play.google.com/store/apps/details?id=jp.gr.java_conf.coskx.ddreader2.trial&hl=en_US&gl=US

**Comment by vincent-karuri** [comment]
I'll try 1

**Comment by f-odhiambo** [comment]
Here is a link to all the images to related lighting conditions while reading the G6PD device https://drive.google.com/drive/folders/1TcdC46RpUBtaAygfDclFSUrLon2QYI8m?usp=sharing

**Comment by vincent-karuri** [comment]
thanks @f-odhiambo something to note here is at some point we need to label the images with the actual readings on them if we go the route of training our own model.

**Comment by vincent-karuri** [comment]
something to keep in mind when installing keras/tensorflow https://stackoverflow.com/a/63243282/9782187

**Comment by vincent-karuri** [comment]
https://stackoverflow.com/a/53422836/9782187

**Comment by vincent-karuri** [comment]
https://stackoverflow.com/a/56994817/9782187 related to tf session config.

**Comment by shoaibmushtaq25** [comment]
So far, no luck for source code or open source library of Yomzo (number 6)

**Comment by vincent-karuri** [comment]
@shoaibmushtaq25 is there information on how 6 can be launched from another app and to get its results after capture?

**Comment by vincent-karuri** [comment]
-  6 seems to perform well on the readings https://github.com/OpenSRP/fhircore/issues/82#issuecomment-821078171. It's however not opensource and there's no clear information on how we can launch it from another app and get the extracted readings. Next steps here would be:
    - Find out if we can get a library (APIs) from the developer (preferred)
    - Alternatively, find out if we can get information about launching the app and getting a reading from it
    
 - 8 has a lot of issues building (is outdated and doesn't have a `requirements.txt`). I could not get it to successfully run and so couldn't evaluate it. Next steps here would be:
   - Try see if we can contact the developers for some guidance 
   - Dedicate time to try and upgrade the outdated parts, test and evaluate
  - 1 takes quite a while to train (been running for over 3 hrs) but for their dataset, it seems that each epoch provides 
 better results. Next steps here would be:
       - train the model
       - crop out G6PD readings and add them to a test folder to see how it performs
       - If above is good OK, otherwise we'd need to train it with labelled images and see if the results are better 
       - It also doesn't seem to be trained on capturing the entire floating point number (from the README images)
  - I haven't gone through 2 and 4 but it looks like we'd need to know how to work with Tessaract to train a model with the data? In that case, next steps are:
     - We'd need to invest time to understand how this works and then evaluate performance.
  - 9 has an interesting approach to encoding the 7-segment display as bits and is a single script which looks simpler. It however only works for that particular display but from skimming it looks as though it's because you have to tell it where to look for the digits (might be a small tweak but not sure). Next steps:
     -  see if we can tweak where it looks for digits and evaluate performance


In most if not all of these, we want to be able to successfully run and subsequently evaluate whether the out-of-box solution works or not. I'm not sure we're a position to know exactly what to tweak right now or after the evaluation but at the least we can try out the solution as is.

**Comment by vincent-karuri** [comment]
![loss_model](https://user-images.githubusercontent.com/33488286/115056561-aeacde80-9eeb-11eb-8096-d3396759180b.png)
![model_as_is](https://user-images.githubusercontent.com/33488286/115056570-b076a200-9eeb-11eb-8bd0-7e1832cd6130.png)

results for 1 after running it.

**Comment by vincent-karuri** [comment]
![loss_model](https://user-images.githubusercontent.com/33488286/115056711-d7cd6f00-9eeb-11eb-94ea-1a9377f00c5e.png)
![model_as_is](https://user-images.githubusercontent.com/33488286/115056717-d8fe9c00-9eeb-11eb-8b19-423ee2023f66.png)

results of running `Model_Single()` from 1 on their training set.

**Comment by pld** [comment]
Cool, let's spend Monday to finish looking into (7), which I think is based on (2).

For (2) yes we'd either need to train of find a model, there should be a model for (2) in https://github.com/arundil/OCR7Segments (ie 7)

**Comment by shoaibmushtaq25** [comment]
6) For number 6, I tried to find out source code or open source libraries used in it but no luck and what I did this morning is that I sent out an email to the email-id mentioned on the Google Play store link of this app "Yomzo" but still there is no reply from them.

7) For number 7, I invested a reasonable time to convert the old eclipse project to android studio project and also converted it to androidx imports and was able to install this app successfully in the mobile. When I run this app, this asks for openCV manager package to install in the mobile. I tried many different ways (Direct from the app or separate installation through terminal) to install the OpenCV manager package and finally I was able to install it few times before.
After installing OpenCV manager package, I run the app but the app is not detecting anything and only showing a rectangle/square of ROI(region of interest). The app also crashes off and on.
here is the screenshot of number 7 app.
![Screenshot_20210419-152528_OCR 7 Segments (1)](https://user-images.githubusercontent.com/12672919/115222740-7ba15f80-a124-11eb-9059-4c987b668d9b.jpg)

**Comment by pld** [comment]
Let's continue looking 7 for another day, don't worry about the crashing, let's see if we can get any of the text rec to work

**Comment by shoaibmushtaq25** [comment]
As I mentioned in the [comment here]([url](https://github.com/OpenSRP/fhircore/issues/82#issuecomment-822364346)) that I sent out an email to developer of "Yomzo" numeral 7-Segment app to ask about the source code or open source library or any way to launch "Yomzo" app from external app. I just received a reply from them. See image below for the details..I will paste it in the github issue as well for the record. 
![Screenshot_20210419-220609_Gmail](https://user-images.githubusercontent.com/12672919/115342673-0be2b180-a1c4-11eb-9ea5-67c9758ae669.jpg)

**Comment by pld** [comment]
@tanvirshanto what was the result of additional research into 7?

**Comment by shoaibmushtaq25** [comment]
@pld I invested yesterday on number 7 and tried to make it work after resolving some outdated code but the app still not working correctly and also crashes off and on while scanning readings through camera. 
I configured NDK and some other required things to make it work but unfortunately no luck.

**Comment by pld** [comment]
Yes of course, K that's helpful thank you, do you see any value in spending more time working on this?

**Comment by shoaibmushtaq25** [comment]
I spent time today as well but still there are some errors coming out due to which app is not working as expected.
I am not sure what would be the next steps now onwards.

**Comment by shoaibmushtaq25** [comment]
After hours of effort, finally I managed to add the native C/C++ code in the final apk of number 7 and then tried to scan the G6PD device readings with it but unfortunately the app doesn't work as expected and crashes too frequently. Looks like the code is too old and not working as expected. I don't think to invest more time on it would be beneficial or give any good results.
cc: @pld @f-odhiambo @vincent-karuri 
![Screenshot from 2021-04-27 16-11-03](https://user-images.githubusercontent.com/12672919/116238676-1defc180-a77b-11eb-8917-6e64a8aa3989.png)

**Comment by pld** [comment]
k thanks @shoaibmushtaq25 for your efforts, looks like we've definitely hit the end of road wrt options we listed here

**Comment by vincent-karuri** [comment]
https://github.com/arturaugusto/display_ocr @shoaibmushtaq25 for this, do you have the script with your changes made  but which was failing?

**Comment by shoaibmushtaq25** [comment]
Yeah, I have its requirement.txt file which I get it prepared from my friend having expertise in python.

**Comment by vincent-karuri** [comment]
a few others that may be interesting:

10. https://github.com/kk5190/Reader-Of-Appliance-Display (Android)
11.  https://github.com/NatholBMX/ScaleDigitizer looks easy to set up (python)
12. https://github.com/arturaugusto/display_ocr have not been able to run this (python)
13. https://www.unix-ag.uni-kl.de/~auerswal/ssocr/ (terminal)
14. https://github.com/auerswal/ssocr github code for above
15. https://github.com/FANMixco/7-segment-ocr-reader (C# inspired by above)
16.  https://github.com/DevashishPrasad/LCD-OCR single script (python)
17. https://github.com/suyashkumar/seven-segment-ocr (python)
18. https://github.com/tvwerkhoven/heat_meter_reader looks specific to heat meters? (python)

**Comment by shoaibmushtaq25** [comment]
Here are the steps and requirements.txt file to run  https://github.com/arturaugusto/display_ocr

1. Install Python 3.8 if not installed yet
2. Add Input Image path that you want to process on `line 64` of file `py3_raspberry_pi.py`
3. Create a `venev` using following command   `python3 -m venv venv_ocr`
4. The activate the venv using command  `source venv_ocr/bin/activate`
5. Then install `requirements.txt`  file given below
6. Then enter command `python py3_raspberry_pi.py`  and see the results

[requirements.txt](https://github.com/opensrp/fhircore/files/6385299/requirements.txt)

**Comment by shoaibmushtaq25** [comment]
I tried number 10 from the list above (https://github.com/kk5190/Reader-Of-Appliance-Display (Android))
And got some reasonable results out of it. I have attached videos of the results. Have a look at the drive [link here](https://drive.google.com/drive/folders/1nAbakQqYeBhINvq1YqULpKLlsY3GPh2M?usp=sharing) to check the accuracy of results with this android project. 
cc: @pld @vincent-karuri @f-odhiambo

**Comment by vincent-karuri** [comment]
> I tried number 10 from the list above (https://github.com/kk5190/Reader-Of-Appliance-Display (Android))
> And got some reasonable results out of it. I have attached videos of the results. Have a look at the drive [link here](https://drive.google.com/drive/folders/1nAbakQqYeBhINvq1YqULpKLlsY3GPh2M?usp=sharing) to check the accuracy of results with this android project.
> cc: @pld @vincent-karuri @f-odhiambo

is there a reading that's displayed somewhere, the [docs](http://kk5190.github.io/Reader-Of-Appliance-Display/) say that it will voice out the reading once it gets it?

**Comment by vincent-karuri** [comment]
Also @shoaibmushtaq25 could you try it out on a mobile phone?

**Comment by shoaibmushtaq25** [comment]
Okay, I will give it a try on mobile phone as well.

**Comment by shoaibmushtaq25** [comment]
> is there a reading that's displayed somewhere, the [docs](http://kk5190.github.io/Reader-Of-Appliance-Display/) say that it will voice out the reading once it gets it?

No, the reading shows only in the ROI(Region of Interest) shown above the reading. Also there is no voice as i observed the app.

**Comment by shoaibmushtaq25** [comment]
> Also @shoaibmushtaq25 could you try it out on a mobile phone?

I am done with the research. I mostly tried it on tablet yesterday and made the video from my mobile phone. and as @vincent-karuri said to test it out on mobile phone as well. so today, I also tried it out on my mobile phone and results are the same as on tablet. 
But still looks like this app is also crashing sometime specially on tapping of a Button given on the screen. I didn't dig into the code yet to see the reason of crashing anyways.
By the way, looks like this can also help us understand to build our solution in the right direction at-least.

**Comment by shoaibmushtaq25** [comment]
so I just reproduced the crash on the number 10 while tapping on a button given on the screen and came to know that this is also throwing the same error `java.lang.UnsatisfiedLinkError` and its obvious reason is not including the native C/C++ code in the final apk. Let me try quickly if I can make it work in this way.

**Comment by shoaibmushtaq25** [comment]
Anyways, I manged to resolve the error around native C/C++ code to add into the apk. 

I was running the app on my mobile phone (android version 11) and openCV package manager was not installing successfully. Then I tried it on my tablet (android version 9) and I manged to install openCV package manager in it successfully. 

App was giving the same results as I shared in the videos attached [here](https://drive.google.com/drive/folders/1nAbakQqYeBhINvq1YqULpKLlsY3GPh2M) .
This app is also crashing off and on. So just like number 7, this number 10 does need to invest some more time to resolve all the old code issues.

**Comment by shoaibmushtaq25** [comment]
I also found a stable apk of number 7 somewhere on the internet and tried it on my device but their results accuracy were not so good. Here is the [link](https://m.apkpure.com/ocr-7-segments/org.ocr.sevensegments.mainActivity) of stable apk of number 7 found on the internet.

**Comment by pld** [comment]
Thanks for the updates, please add related code here for now, https://github.com/opensrp/diagnostics/tree/main/g6pd

**Comment by shoaibmushtaq25** [comment]
@pld , Related code of number 10 is uploaded to the above url https://github.com/opensrp/diagnostics/tree/main/g6pd

**Comment by f-odhiambo** [comment]
Research summary 

1. Reviewed APK/Code
Paid version  ($2) version of Yomzo app plus the [documentation](http://coskx.webcrow.jp/mrr/yomzoGuide3/guide_en.html )  - No improvement 
- Optical Character Recognition of Seven Segment Display (Python based) - [link](https://github.com/NickTrossa/OCR7SD )
- [link](https://github.com/SachaIZADI/Seven-Segment-OCR )
- Seven segment libs  - [links](https://github.com/topics/seven-segment?o=desc&s=forks )
- optical-character-recognition - [link](https://github.com/topics/optical-character-recognition?l=c&o=asc&s=stars&utf8=%E2%9C%93 )
- tesseract-ocr [link](https://github.com/tesseract-ocr/)
- Review Google AutoML framework [link](https://cloud.google.com/automl) . Also contacted Fred, who referred us to Thye Yeow Bok, ML Kit expert to try to help us with the research 


2. Compiled list of literature read
- Automated method for detecting and reading seven-segment digits from images of blood glucose metres and blood pressure monitors - [link](https://www.tandfonline.com/doi/full/10.1080/03091902.2019.1673844 )  - Fred has some leads already connected with him and will be sharing contacts soon 
- Real-Time Detection and Reading of LED/LCD Displays for Visually Impaired Persons [link](https://www.ski.org/sites/default/files/publications/wacv11-display-reader.pdf )
- Seven Segment Optical Character Recognition [link](https://www.unix-ag.uni-kl.de/~auerswal/ssocr/ )
- Utilizing Smartphone-Based Machine Learning in Medical Monitor Data Collection: Seven Segment Digit Recognition  - [link](https://arxiv.org/pdf/1807.04888.pdf ) - closest research work to G6PD
- How to recognize Seven Segment Displays content with ssocr (Seven Segment Optical Character Recognition) in Ubuntu 16.04 - [link](https://ourcodeworld.com/articles/read/741/how-to-recognize-seven-segment-displays-content-with-ssocr-seven-segment-optical-character-recognition-in-ubuntu-1604 ) can be server-based 
- Text detection and recognition in raw image dataset of seven-segment digital energy meter display - [link](https://www.sciencedirect.com/science/article/pii/S235248471930174X )
- Measurement Data Recognition from SevenSegment Indicator by Mobile Device - [link](https://www.fruct.org/sites/default/files/files/conference15/timofeev.pdf ) This is good research on current problems we are facing

**Comment by shoaibmushtaq25** [comment]
Here is the link to more demo videos [link](https://drive.google.com/drive/folders/1s0M82rRr1ywmJESfyrSdyX67S-qNtmbG?usp=sharing) of results with 

- Google ML Kit
- Number 7 (https://github.com/arundil/OCR7Segments/tree/master/OCR7SegmentsApp)
- Number 10 (https://github.com/kk5190/Reader-Of-Appliance-Display)

**Comment by shoaibmushtaq25** [comment]
Installed and run desktop version of Tesseract OCR and tested it with different Input Images , i.e, English Text, Numeric Text, G6PD device reading Image, Cropped out seven segment digits and results are documented [here](https://docs.google.com/document/d/1lSe8QM4YNgzgnmYk74WMoZY58gCO5dGprHomK5J633c/edit?usp=sharing)

**Comment by vincent-karuri** [comment]
SSOCR initial findings: https://docs.google.com/document/d/1RAPfoZhV5koVR-HUrftMoe-Zs7YcfPfqljc35HJfUJA/edit

**Comment by shoaibmushtaq25** [comment]
Tried already available seven segment display trained data sets with desktop based Tesseract OCR at different resolutions . Results are [here](https://docs.google.com/document/d/1lSe8QM4YNgzgnmYk74WMoZY58gCO5dGprHomK5J633c/edit)
